Slim full builds

guyush1 opened the pull request.

guyush1 committed:
build: split build into 2 variants - slim & full

Full builds also include python (similar to -with-python before) & also
include cross architecture debugging. full builds are configurable via
the full_build_conf.sh file.

guyush1 committed:
docs: Updated README.md with new build instructions

guyush1 committed:
cicd: Updated github workflow with the new makefile rules

guyush1 committed:
build: disable simulator when building with full_build_cross_arch_debugging

Since not all targets have a supported simulator, there were compilation
errors. The configure file also appears to be half-broken, so we pass
--disable-sim when this feature is enable to avoid compilation failiure.

roddyrap commented on Makefile:
```diff
@@ -1,12 +1,17 @@
-ALL_PACK_TARGETS := $(PACK_TARGETS) $(PYTHON_PACK_TARGETS)
+SLIM_TARGETS := $(addsuffix -slim, $(BASE_TARGETS))
+FULL_TARGETS := $(addsuffix -full, $(BASE_TARGETS))
+ALL_TARGETS := $(SLIM_TARGETS) $(FULL_TARGETS)
```
Maybe add BUILD to the names? Having ALL_TARGETS that doesn't contain pack targets is strange.

guyush1 commented:
CICD has passed. I just pushed some documentation changes now, and do not want to rerun our cicd for them.

roddyrap commented on README.md:
```diff
@@ -64,8 +64,10 @@ To get started with `gdb-static`, simply download the build for your architectur
+> 1. Full builds with Python support && cross-architecture debugging, ~70MB in size. Note that we disable the simulator feature in order to enable cross-architecture debugging (since not all targets have a simulator).
+> 2. Slim builds, which are approximately ~10 MB in size.
+>
+> Full builds are modular - you can edit full_build_conf.sh to disable parts of the full builds.
```
Include the possible modifications.

roddyrap commented on README.md:
````diff
@@ -97,10 +99,10 @@ Building for a specific architecture
 ```
 
-Where `<ARCH>` is the architecture you want to build for, and `-with-python` may be added in order to compile gdb with Python support.
+Where `<ARCH>` is the architecture you want to build for, and `slim/full` determines the build type (see note about builds above).
````
Link to the relevant title.

roddyrap commented on src/compilation/build.sh:
```diff
@@ -3,6 +3,7 @@
 # Include utils library
 script_dir=$(dirname "$0")
 source "$script_dir/utils.sh"
+source "$script_dir/full_build_conf.sh"
```
Make this an optional environment variable.
```suggestion
source "${GDB_STATIC_BUILD_CONF_FILE:-$script_dir/full_build_conf.sh}"
```

guyush1 replied:
Not doing this right now, since in order for this to work, we need to pass the env variable into the docker, and our build system is cluttered as is. we can support it when we re-write the build system later on :)

roddyrap commented on src/compilation/build.sh:
```diff
@@ -489,10 +507,12 @@ function build_gdb() {
+    # $6: liblzma prefix
+    # $7: whether to build gdb with all extra configurations specified in src/compilation/full_build_conf.sh
+    # $8: gdb cross-architecture binary format support formats (relevant for full builds only).
+    # $9: the architecture we are compiling for.
```
Put 9 before 8 because it's always needed.

roddyrap commented on src/compilation/build.sh:
```diff
@@ -489,10 +507,12 @@ function build_gdb() {
+    # $4: libgmp prefix
+    # $5: libmpfr prefix
+    # $6: liblzma prefix
+    # $7: whether to build gdb with all extra configurations specified in src/compilation/full_build_conf.sh
```
Mention slim/full

roddyrap commented on src/compilation/build.sh:
```diff
@@ -504,17 +524,29 @@ function build_gdb() {
-        local python_flag="--without-python"
-        local gdb_build_dir="$(realpath "$gdb_dir/build-${target_arch}")"
+        extra_flags+=("--without-python")
+        local gdb_build_dir="$(realpath "$gdb_dir/build-${target_arch}_slim")"
```
```suggestion
        local gdb_build_dir="$(realpath "$gdb_dir/build-${target_arch}-slim")"
```

This is how you called the build alias, and I want them to be compatible.

guyush1 committed:
build system improvements:

	I) Better names for makefile variables.
	II) Improved documentation on build types in the md files.
	III) Removed redundant parameters in build.sh & Improved
internal documentation
